include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -902,124 +842,265 @@ void EndParallelSectionInternal(PerThread &pt,
+    }
+  }
+
+  // ...and wait for any other tasks not revoked to finish their work
+  auto tasks_to_wait_for = tasks_started - ps.tasks_revoked;
   while (ps.tasks_finished < tasks_to_wait_for) {
     onnxruntime::concurrency::SpinPause();
   }
+  
   // Clear status to allow the ThreadPoolParallelSection to be
   // re-used.
   ps.tasks_finished = 0;
-
-  if (ps.dispatch_q_idx > -1) {
-    // if dispatch task started, wait for its work completion
-    while (!ps.work_done.load(std::memory_order_acquire)) {
-      onnxruntime::concurrency::SpinPause();
-    }
-  }
 }
 
 void EndParallelSection(ThreadPoolParallelSection &ps) override {
   PerThread* pt = GetPerThread();
   EndParallelSectionInternal(*pt, ps);
 }
 
-// RunInParallelInternal dispatch tasks to a number of workers asynchronously.
-// A worker thread will be selected as the dispatcher that distributes tasks.
+//----------------------------------------------------------------------
+//
+// Preferred workers
+// -----------------
+//
+// Initialize the set of hints for preferred worker threads we will
+// use.  We do this once, covering the maximum num_threads_ items,
+// in order to avoid resizing preferred_workers concurrent with
+// access from worker threads.
+//
+// For simplicity we initialize with hints round-robin among the
+// workers.  For simple workloads with 1 main thread this means we
+// will distribute work across the pool of workers.  For workers
+// with multiple main threads it attempts to balance the load.
+//
+// These hints are just used as a starting point, and are updated by
+// the worker thread that actually claims an item (e.g., if an item
+// initially assigned to thread T1 is stolen and executed by T2,
+// then T2 is assigned at the new preferred worker).
+//
+// Note that the hints are held in the _main_ thread that submits
+// work to the pool.  We assume that a thread is primarily
+// submitting work to just one pool, but allow for the pool to
+// change over time.  Hence we allow the hints vector to grow over
+// time.
+//
+//
+// A note on terminology used in the variable names here:
+//
+// dop - degree of parallelism, as seen by the user.  For instance
+//       dop=4 means 4 threads in total: 1 main thread that enters the
+//       loop, plus 1 dispatcher thread, plus 2 additional worker
+//       threads.
+//
+// par_idx - a thread's index within the loop, in the range [0,dop).
+//
+// num_threads_ - the number of worker threads in the thread pool.  A
+//       loop with dop=4 will be common on a pool with 3 threads
+//       (given that the main thread will also participate).
+//
+// q_idx - a worker queue index, in the range [0,num_threads_).
+ 
+void InitializePreferredWorkers(std::vector<int> &preferred_workers) {
+  static std::atomic<unsigned> next_worker;
+  // preferred_workers maps from a par_idx to a q_idx, hence we
+  // initialize slots in the range [0,num_threads_]
+  while (preferred_workers.size() <= num_threads_) {
```

Comment:
>num_threads_ [](http://example.com/codeflow?start=37&length=12)

Why not just return next_worker.fetch_add(n, std::memory_order_relaxed) as begining thread, and let the caller fill its own array? 
This way would save (n - 1) overhead on increasing a static atomic brought by the while loop?

Reply:
This is just aiming for simplicity here -- it will execute once per main thread at start-up, and then only again if the same main thread uses a different thread pool with more threads.  We could avoid some atomic CAS/FADD by claiming a set of IDs together, but the logic would get more complex.